lib/kubeclient/informer.rb

```diff
@@ -78,37 +72,53 @@ def fill_cache
     end
 
     def watch_to_update_cache
+      watcher_with_timeout do |watcher|
+        stop_reason = 'disconnect'
+
+        watcher.each do |notice|
+          case notice[:type]
+          when 'ADDED', 'MODIFIED' then @cache[cache_key(notice[:object])] = notice[:object]
+          when 'DELETED' then @cache.delete(cache_key(notice[:object]))
+          when 'ERROR'
+            stop_reason = 'error'
+            break
+          else
+            @logger&.error("Unsupported event type #{notice[:type]}")
+          end
+          @watching.each { |q| q << notice }
+        end
+
+        stop_reason
+      end
+    end
+
+    def watcher_with_timeout
       @watcher = @client.watch_entities(@resource_name, watch: true, resource_version: @started)
-      stop_reason = 'disconnect'
+      timeout_deadline = Time.now + @reconcile_timeout
+      watcher_finished = false
 
-      # stop watcher without using timeout
-      @waiter = Thread.new do
-        sleep(@reconcile_timeout)
-        stop_reason = 'reconcile'
+      finisher_thread = Thread.new do
+        sleep(0.5) until @stopped || watcher_finished || Time.now > timeout_deadline
+        # loop calling finish until the actual method has
+        # exited, since watcher.each may be called after the
+        # finish in this thread is called
         @watcher.finish
+        @watcher.finish until watcher_finished
```

Comment:
might want to add a small delay here

Reply:
Indeed, might even let the watch thread exit faster than busy-waiting?  Even if not, it's polite to other threads & processes to limit how much CPU this can eat.